core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java

```diff
@@ -314,14 +315,36 @@ public void onComplete() {
                 responseEncoder.writeTrailers(req.id(), req.streamId(), additionalTrailers)
                                .addListener(writeHeadersFutureListener(true));
                 ctx.flush();
-            } else if (responseEncoder.isWritable(req.id(), req.streamId())) {
-                responseEncoder.writeData(req.id(), req.streamId(), HttpData.empty(), true)
-                               .addListener(writeDataFutureListener(true, true));
-                ctx.flush();
+            } else {
+                if (isWritable()) {
+                    responseEncoder.writeData(req.id(), req.streamId(), HttpData.empty(), true)
+                                   .addListener(writeDataFutureListener(true, true));
+                    ctx.flush();
+                } else {
+                    if (!reqCtx.sessionProtocol().isMultiplex()) {
+                        // An HTTP/1 connection is closed by a remote peer after all data is sent,
+                        // so we can assume the HTTP/1 request is complete successfully.
+                        succeed();
+                    } else {
+                        fail(ClosedStreamException.get());
```

Comment:
question: Should we treat HTTP/2 connection when `onComplete` is called if it isn't writable?

Reply:
It would be an invalid state for HTTP/2 because connections are gracefully shutting down with GOAWAY.

Reply:
Got it. Thanks!